debugger/src/debugger.rs

```diff
@@ -403,6 +408,76 @@ impl DebugSession {
     }
 }
 
+/// Manage the active RTT target for a specific DebugSession, as well as provide methods to reliably move RTT from target, through the debug_adapter, to the client.
+pub(crate) struct DebuggerRttTarget {
+    /// The  connection to RTT on the target
+    target_rtt: rtt::RttActiveTarget,
+    /// Some status fields and methods to ensure conitnuity in flow of data from target to debugger to client.
+    debugger_rtt_channels: Vec<DebuggerRttChannel>,
+}
+
+impl DebuggerRttTarget {
+    pub fn process_rtt_data<P: ProtocolAdapter>(
+        &mut self,
+        debug_adapter: &mut DebugAdapter<P>,
+        target_core: &mut Core,
+    ) -> bool {
+        let mut at_least_one_channel_had_data = false;
+        for debugger_rtt_channel in self.debugger_rtt_channels.iter_mut() {
+            if debugger_rtt_channel.send_rtt_data(target_core, debug_adapter, &mut self.target_rtt)
+            {
+                at_least_one_channel_had_data = true;
+            }
+        }
+        at_least_one_channel_had_data
+    }
+}
+
+pub(crate) struct DebuggerRttChannel {
+    pub(crate) channel_number: usize,
+    // We will not poll target RTT channels until we have confirmation from the client that the output window has been opened.
+    pub(crate) has_client_window: bool,
+}
+impl DebuggerRttChannel {
+    /// Poll and retrieve data from the target, and send it to the client, depending on the state of `hasClientWindow`.
+    /// Doing this selectively ensures that we don't pull data from target buffers until we have a output window, and also helps us drain buffers after the target has entered a `is_halted` state.
+    /// Errors will be reported back to the `debug_adapter`, and the return `bool` value indicates whether there was available data that was processed.
+    pub(crate) fn send_rtt_data<P: ProtocolAdapter>(
+        &mut self,
+        core: &mut Core,
+        debug_adapter: &mut DebugAdapter<P>,
+        rtt_target: &mut rtt::RttActiveTarget,
+    ) -> bool {
+        if self.has_client_window {
+            rtt_target
+                .active_channels
+                .iter_mut()
+                .find(|active_channel| {
+                    if let Some(channel_number) = active_channel.number() {
+                        channel_number == self.channel_number
+                    } else {
+                        false
+                    }
+                })
+                .and_then(|rtt_channel| {
+                    rtt_channel.get_rtt_data(core, rtt_target.defmt_state.as_ref())
+                })
+                .and_then(|(channel_number, channel_data)| {
+                    if debug_adapter
+                        .rtt_output(channel_number.parse::<usize>().unwrap_or(0), channel_data)
```

Comment:
This doesn't fome from a structure, but from the `get_rtt_data()` method of `RttActiveChannel`. I created it based on the method of the same name on `RttActiveTarget`.  I chose not to modify that API, because I didn't want to negatively impact other consumers. 

I figured it was highly unlikely that it would pass back a channel number that wouldn't parse, so this `unwrap_or` is simply because I have an aversion for panicking `unwrap` calls. The worst that could happen is it would send data from non_zero channels to the zero channel.

That said, I'm happy to make the change to the API if you think that is better in the long run.

Reply:
Yeah, I think we should change the API think this would be better to be changed :)
Validation should imo always be performed upfront at the earliest stage possible and validated values should be used in further code. That is one of the strenghts of a typesystem and makes complicated internal code easier to argue about :)

Reply:
Ok, let's fix this another time. After all this problem is some place else in the API of `probe-rs-cli-util`.